ntp-proto/src/packet.rs

```diff
@@ -140,41 +269,135 @@ impl<'a> ExtensionField<'a> {
+        use PacketParsingError::IncorrectLength;
+
+        if data.len() < 4 || data.len() % 4 != 0 {
+            return Err(IncorrectLength);
         }
+
         let typeid = u16::from_be_bytes(data[0..2].try_into().unwrap());
         let ef_len = u16::from_be_bytes(data[2..4].try_into().unwrap()) as usize;
-        if ef_len < Self::MINIMUM_SIZE || ef_len > data.len() {
+        if ef_len < Self::MINIMUM_SIZE {
             return Err(PacketParsingError::IncorrectLength);
         }
-        Ok((
-            ExtensionField::Unknown {
+
+        let value = data.get(4..ef_len).ok_or(IncorrectLength)?;
+
+        // check that the padding is all zeros. This is required for the fuzz tests to work
+        let padding = &data[ef_len..next_multiple_of(ef_len, 4)];
+        if padding.iter().any(|b| *b != 0) {
+            return Err(PacketParsingError::IncorrectLength);
+        }
+
+        let field = match typeid {
+            0x104 => {
+                // The string MUST be at least 32 octets long
+                if value.len() < 32 {
+                    return Err(IncorrectLength);
+                }
+
+                ExtensionField::UniqueIdentifier(value[..].into())
+            }
+            0x204 => ExtensionField::NtsCookie(value[..].into()),
+            0x304 => {
+                if value.iter().any(|b| *b != 0) {
+                    return Err(PacketParsingError::IncorrectLength);
+                }
+
+                ExtensionField::NtsCookiePlaceholder {
+                    body_length: value.len() as u16,
+                }
+            }
+            0x404 => {
+                let nonce_length = u16::from_be_bytes(value[0..2].try_into().unwrap()) as usize;
+                let ciphertext_length =
+                    u16::from_be_bytes(value[2..4].try_into().unwrap()) as usize;
+
+                if 4 + next_multiple_of(nonce_length, 4) + next_multiple_of(ciphertext_length, 4)
+                    != next_multiple_of(value.len(), 4)
+                {
+                    return Err(PacketParsingError::IncorrectLength);
+                }
+
+                let ciphertext_start = 4 + next_multiple_of(nonce_length as usize, 4);
+
+                let nonce = value.get(4..4 + nonce_length).ok_or(IncorrectLength)?;
+                let nonce_padding = value
+                    .get(4 + nonce_length..ciphertext_start)
+                    .ok_or(IncorrectLength)?;
+
+                if nonce_padding.iter().any(|b| *b != 0) {
+                    return Err(PacketParsingError::IncorrectLength);
+                }
+
+                let ciphertext = value
+                    .get(ciphertext_start..ciphertext_start + ciphertext_length)
+                    .ok_or(IncorrectLength)?;
+
+                let plaintext = match cipher.decrypt(Nonce::from_slice(nonce), ciphertext) {
+                    Ok(plain) => plain,
+                    Err(e) => panic!("{e:?}"),
+                };
```

Comment:
this allocates. Doing the decryption in-place does not work with the rest of the code (because the data behind the mutable reference is then not available for further decoding. 

We could pass in a "scratchpad" buffer in the future (using some sort of pool allocator) to minimize the allocation cost. I'm not sure that's worth it, but it's an option.